lib/spack/spack/binary_distribution.py

```diff
@@ -266,10 +267,7 @@ def find_by_hash(self, find_hash, mirrors_to_check=None):
                 None, just assumes all configured mirrors.
         """
         if find_hash not in self._mirrors_for_spec:
-            # Not found in the cached index, pull the latest from the server.
-            self.update(with_cooldown=True)
```

Comment:
Can you check if by now it's even a significant overhead?

Reply:
I was going to take a minute today to run this PR in our CI, but it fell over immediately with an `Access Denied` error from `spack ci generate`. Haven't investigated yet, but @haampie does https://github.com/spack/spack/pull/34324 have something to do with this? Last I saw this failure was from before https://github.com/spack/spack/pull/31391 was merged.

I'll try again once things have settled a bit, it's hard to replicate our exact CI environment and there are other tasks (and Spack PRs!) that need my attention.

> I don't think it does, it's a different code path, and it ensures the local cache is up to date

My point is that folks now assume `spack spec --reuse zlib` will always give the latest `zlib` in their enabled buildcaches. If the `index.json` is only fetched by explicit request (e.g. `spack buildcache fetch`), that behavior will be wildly different.

Reply:
> My point is that folks now assume spack spec --reuse zlib will always give the latest zlib in their enabled buildcaches. If the index.json is only fetched by explicit request (e.g. spack buildcache fetch), that behavior will be wildly different.

That behavior is not going to change. `spack spec` requires an up-to-date index.json, so it will fetch it. `spack install` will not, but given that people concretize with reuse and then fetch, the cache is typically up to date.

#34324 was reverted (again) :weary: does `head_object` give you access denied or so? I think https://github.com/spack/spack/commit/7e054cb7fc55c35d50480f04552536eed48f5e21 is more likely to be the issue.